### PR TITLE
ORM, Fields, Validation

### PR DESCRIPTION
**Pull request by dancio**

### CODE REVIEW

**Comment by atteroTheGreatest** [comment]
@dancio what's wrong with the tests? Why don't they pass?

File: syncano/__init__.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 __copyright__ = 'Copyright 2015 Syncano'
 
 VERSION = __version__
-API_ROOT = 'http://127.0.0.1:8000/api/'
+API_ROOT = os.getenv('SYNCANO_API_ROOT', 'https://api.syncano.com/')
```
Comment (atteroTheGreatest) [inline syncano/__init__.py]: You can use https://v4.hydraengine.com if you like ;).

File: syncano/models/base.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        self.validate()
+        data = self.to_native()
+
+        if self.links:
+            endpoint = self.links['self']
+        else:
+            endpoint = self._meta.resolve_endpoint('list', data)
+
+        request = {'data': data}
+        response = self._meta.connection.request('POST', endpoint, **request)
```
Comment (atteroTheGreatest) [inline syncano/models/base.py]: Shouldn't it be put/patch if model's instance already exists?

Comment (dancio) [inline syncano/models/base.py]: good catch +1 in this case `PUT`

File: syncano/models/fields.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    def validate(self, value, model_instance):
+        if self.required and not value:
+            raise self.VaidationError('This field is required.')
+
+        if self.read_only and getattr(model_instance, self.name):
+            raise self.VaidationError('Field is read only.')
+
+        if isinstance(value, six.string_types):
+            if self.max_length and len(value) > self.max_length:
+                raise self.VaidationError('Max length reached')
```
Comment (atteroTheGreatest) [inline syncano/models/fields.py]: Period.

File: syncano/models/fields.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+            if self.min_length and len(value) < self.min_length:
+                raise self.VaidationError('Max length reached')
```
Comment (atteroTheGreatest) [inline syncano/models/fields.py]: Period.

File: syncano/models/fields.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+class IntegerField(WritableField):
+
+    def to_python(self, value):
+        if value is None:
+            return value
+        try:
+            return int(value)
+        except (TypeError, ValueError):
+            raise self.VaidationError('Invalid value.')
```
Comment (atteroTheGreatest) [inline syncano/models/fields.py]: Maybe information about expected type would be helpful? For example: `'Invalid value. Value should be integer.'`

File: syncano/models/fields.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+            return value
+
+        try:
+            parsed = self.parse_date(value)
+            if parsed is not None:
+                return parsed
+        except ValueError:
+            pass
+
+        raise self.VaidationError('Invalid date.')
```
Comment (atteroTheGreatest) [inline syncano/models/fields.py]: Add information about date format here. It's not obvious for everyone. Especially Europeans vs Americans ;).

File: syncano/models/fields.py [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+            pass
+
+        try:
+            parsed = self.parse_date(value)
+            if parsed is not None:
+                return datetime(parsed.year, parsed.month, parsed.day)
+        except ValueError:
+            pass
+
+        raise self.VaidationError('Invalid value.')
```
Comment (atteroTheGreatest) [inline syncano/models/fields.py]: Add information about expected format.

File: syncano/models/manager.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        return type(self).__bool__(self)
+
+    # Object actions
+
+    def create(self, **kwargs):
+        instance = self.model(**kwargs)
+        instance.save()
+        return instance
+
+    def bulk_create(self, objects):
```
Comment (atteroTheGreatest) [inline syncano/models/manager.py]: It would be cool if we had a real support for bulk creates :).

File: syncano/models/manager.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+
+    def delete(self, *args, **kwargs):
+        self.method = 'DELETE'
+        self.endpoint = 'detail'
+        self._filter(*args, **kwargs)
+        return self.request()
+
+    def update(self, *args, **kwargs):
+        self.method = 'POST'
+        self.endpoint = 'detail'
```
Comment (atteroTheGreatest) [inline syncano/models/manager.py]: Hm... POST to detail?

File: syncano/models/manager.py (hunk 8 above)
Comment (dancio) [inline syncano/models/manager.py]: maybe some day :)

File: syncano/models/manager.py [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+
+    def _filter(self, *args, **kwargs):
+        if args and self.endpoint:
+            properties = self.model._meta.get_endpoint_properties(self.endpoint)
+            mapped_args = {k: v for k, v in zip(properties, args)}
+            self.properties.update(mapped_args)
+        self.properties.update(kwargs)
+
+    def _clone(self):
+        # Maybe deepcopy ?
```
Comment (atteroTheGreatest) [inline syncano/models/manager.py]: Maybe it's maybelline? ;) Deepcopy would definitely make code shorter, but what about memory footprint? We don't need deep copies of everything.

File: syncano/models/manager.py [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+        return manager
+
+    def serialize(self, data):
+        if not isinstance(data, dict):
+            return
+        return self.model(**data) if self._serialize else data
+
+    def request(self, method=None, path=None, **request):
+        meta = self.model._meta
+        method = method or self.method
```
Comment (atteroTheGreatest) [inline syncano/models/manager.py]: I know that using state of the Manager, makes function calls shorten, but uh, I prefer more functional approach.

File: syncano/models/registry.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+    def register_definition(self, definition):
+        Meta = type(str('Meta'), (Options, ), {
+            'connection': self.connection,
+            'endpoints': definition['endpoints'],
+            'name': definition['name'],
+        })
+
+        attrs = {'Meta': Meta}
+        for name, options in six.iteritems(definition.get('properties', {})):
```
Comment (atteroTheGreatest) [inline syncano/models/registry.py]: `options` is a different thing than `Options`, right? Maybe it should have a different name to minimize confusion?

File: syncano/resultset.py [hunk 13]
```diff
@@ -1,35 +0,0 @@
-
-
-class ResultSet(object):
```
Comment (atteroTheGreatest) [inline syncano/resultset.py]: What about result_sets now? Can we iterate over lists and so on?

Comment (dancio) [inline syncano/resultset.py]: yes, whole functionality is now in Manager

File: syncano/models/fields.py [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+            kw = {k: int(v) for k, v in six.iteritems(match.groupdict())}
+            return date(**kw)
+
+    def to_native(self, value):
+        if isinstance(value, datetime.datetime):
+            value = value.date()
+        return value.isoformat()
+
+
+class DateTimeField(DateField):
```
Comment (atteroTheGreatest) [inline syncano/models/fields.py]: What about timezones?

Comment (dancio) [inline syncano/models/fields.py]: I'll add ticket for that currently we don't support them

Comment (23doors) [inline syncano/models/fields.py]: Do you mean the library or the platform? In the platform, all our dates are expected to be converted to UTC timezone and that's what we output as well. We do not plan nor need any further support from platform perspective. But easily enough, library can allow setting default timezones and convert those datetimes on the fly.

**Pull request merged**